src/components/TodoItem.tsx

```diff
@@ -0,0 +1,262 @@
+  } = todo;
+
+  const {
+    todos,
+    shouldLoading,
+    newTodoInputRef,
+  } = useContext(StateContext);
+  const dispatch = useContext(DispatchContext);
+  const [isLoading, setIsLoading] = useState(!todos.find(t => t.id === id));
+  const [isUpdatingTitle, setIsUpdatingTitle] = useState(false);
+  const [newTitle, setNewTitle] = useState(title);
+  const updateTitleRef = useRef<HTMLInputElement | null>(null);
+
+  const toggleChecked = async (e: React.ChangeEvent<HTMLInputElement>) => {
+    try {
+      setIsLoading(true);
+      const updatedTodo = await updateTodo({
+        id,
+        title,
+        completed: e.target.checked,
+        userId,
+      });
+
+      dispatch({
+        type: 'updateTodo',
+        payload: { todo: updatedTodo },
+      });
+    } catch (error) {
+      dispatch({
+        type: 'error',
+        payload: { error: ErrorMessage.Updating },
+      });
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  const deleteCurrentTodo = async () => {
+    try {
+      setIsLoading(true);
+      await deleteTodo(id);
+
+      dispatch({
+        type: 'deleteTodo',
+        payload: { id },
+      });
+    } catch (error) {
+      dispatch({
+        type: 'error',
+        payload: { error: ErrorMessage.Deleting },
+      });
+      setNewTitle(title);
+    } finally {
+      setIsUpdatingTitle(false);
+      setIsLoading(false);
+      newTodoInputRef?.focus();
+    }
+  };
+
+  const handleNewTitleSubmit = async () => {
+    try {
+      setIsLoading(true);
+
+      if (newTitle.trim() === title) {
+        setIsUpdatingTitle(false);
+        setIsLoading(false);
+
+        return;
+      }
+
+      const updatedTodo = {
+        ...todo,
+        title: newTitle.trim(),
+      };
+
+      if (!newTitle.trim()) {
+        try {
+          await deleteTodo(id);
+
+          dispatch({
+            type: 'deleteTodo',
+            payload: { id },
+          });
+        } catch (error) {
+          if (isUpdatingTitle && updateTitleRef.current) {
+            updateTitleRef.current.focus();
+          }
+
+          dispatch({
+            type: 'error',
+            payload: { error: ErrorMessage.Deleting },
+          });
+        } finally {
+          setIsLoading(false);
+        }
+
+        return;
+      }
+
+      await updateTodo(updatedTodo);
+
+      dispatch({
+        type: 'updateTodo',
+        payload: { todo: updatedTodo },
+      });
+      setIsUpdatingTitle(false);
+    } catch (error) {
+      if (isUpdatingTitle && updateTitleRef.current) {
+        updateTitleRef.current.focus();
+      }
+
+      dispatch({
+        type: 'error',
+        payload: { error: ErrorMessage.Updating },
+      });
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  const handleKeyUp = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === 'Escape') {
+      setIsUpdatingTitle(false);
+      setNewTitle(title);
+    }
+  };
+
+  useEffect(() => {
+    switch (true) {
```

Comment:
hm..it's not obvious solution to work with loading like this. you'd better setIsLoading(true/false) during async operations

Reply:
Agree with comment
![Uploading 2023-12-13_15h44_40.png…]()